### ISSUE TITLE
Make `openModal` a promise that resolves when the modal closes

### ISSUE THREAD

**Issue by c00** [mattjennings/svelte-modals#17]
**Is your feature request related to a problem? Please describe.**
Consider a confirmation dialog. A user gets a message "Are you sure?" and clicks yes or no. Currently you'd have to do this with  callback props, they would have to exist on any modal component you'd want to wait for or have a return value for.

This works, but isn't optimal.

**Describe the solution you'd like**
Have the `openModal` function return a promise. This promise will resolve once the modal is closed.

This could be done by extending the `StoredModal` interface with a `promise`, and then on `pop()` just resolve the promises of the modals you're popping.

Then update `closeModal()` to accept a parameter `returnValue?: unknown`. That value will be handed to the eventual `pop` function so it can be used to resolve the Promise. (To be helpful, add a generic to `openModal` so we know what to expect when it resolves)

Now what a user could do is:

```ts
const confirmation = await openModal(MyFancyConfirmationDialog);
if (confirmation) doStuff();
```

And in the modal itself, no extra callbacks or boilerplatey stuff would be required, just calling the close method would be sufficient:

```html
    <button type="button" on:click={() => closeModal(false)}> Cancel </button>
    <button type="button" on:click={() => closeModal(true)}> Confirm </button>
```

**Describe alternatives you've considered**
Rather than returning the promise from `openModal` directly, it could also return an object with the promise, so in the future we could add more stuff to it. Either way is fine.

I'd be very willing to make a PR for this. Just let me know if you'd be willing to accept such a PR.

**Comment by mattjennings** [comment]
I think this would be a good idea. This is likely a common scenario, and I don't see any drawbacks to having `openModal` just be a promise until it closes anyways. Resolving with just the data would be my preference rather than an object containing the data.

A few thoughts:

- What would you expect to happen in the case of `closeAllModals()`? Should it also take a response that would resolve all `openModal` promises with?

- It would be good if there was a way to infer the response type from a type defined in the modal component, rather than having it be unknown. This might be possible through the `$$Generic` type, but I'd have to take a closer look. If you have any ideas here let me know, otherwise I'd be happy to tackle this after you create your PR.

**Comment by mattjennings** [comment]
This will be added in #20 but with a slightly different API. 

Instead of providing the value to `closeModal`, modal components receive a `close` prop which is a function that can be used to close itself and return a value. 

Reasons why:

- it allows the return type to be inferred by the component type
- it tightens the relationship from the modal to its `openModal` call, making sure that the return type is the expected type and not unexpectedly received from `closeModals()` (that may not have been intentional).
    - if `closeModal/s` is called, `openModal` resolves with undefined.

**Comment by c00** [comment]
I like this a lot! Thank you for working on this!

**Comment by mattjennings** [comment]
included in v2.0.0
